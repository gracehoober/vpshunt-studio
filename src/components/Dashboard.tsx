import { NavBar } from "./NavBar";
import { Button, Box } from "@mui/material";
import { useState } from 'react';
import { NewEntry } from "./NewEntry";
import React from "react";


const Dashboard: React.FC = () => {
    const [isNewEntryOpen, setIsNewEntryOpen] = useState(false);
    return (
        <Box sx={styles.container}>
            <NavBar />
            <Button
                variant="contained"
                sx={styles.button}
                onClick={() => setIsNewEntryOpen(true)}>
                Add new entry
            </Button>
            {isNewEntryOpen && <NewEntry
                open={isNewEntryOpen}
                onClose={() => setIsNewEntryOpen(false)} />}
        </Box>
    );
};

const styles = {
    container: {
        backgroundColor: 'background.dashboard',
        width: '100vw',
        height: '100vh',
        padding: 4,
    },
    button: {
        color: 'background.paper',
        p: 1,
        m: 4
    }
};
export { Dashboard };
