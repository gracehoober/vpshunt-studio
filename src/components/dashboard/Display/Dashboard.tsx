import { Button, Box, IconButton, LinearProgress, Stack } from "@mui/material";
import { useState, useEffect, useCallback } from "react";
import { NewEntry } from "../NewEntry/NewEntry";
import React from "react";
import { fetchUserShuntData } from "../../../api/shunts";
import type { DashboardData } from "../types";
import { createColumns } from "../utils";
import type { Rows } from "../types";
import { StyledDashboardGrid } from "./StyledDashboardGrid";
import { useTranslation } from "react-i18next";
import RefreshIcon from "@mui/icons-material/Refresh";
import { ErrorAlert } from "../../common/ErrorAlert";

const Dashboard: React.FC = () => {
  const { t } = useTranslation();
  const [isNewEntryOpen, setIsNewEntryOpen] = useState(false);
  const [dashboardData, setDashboardData] = useState<DashboardData[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  const fetchDashboardData = useCallback(() => {
    setLoading(true);
    setError(null);

    fetchUserShuntData()
      .then((data) => {
        setDashboardData(data);
        setLoading(false);
      })
      .catch((err) => {
        setError(t("dashboard.error"));
        setLoading(false);
        console.error("Failed to fetch dashboard data:", err);
      });
  }, [t]);

  useEffect(() => {
    fetchDashboardData();
  }, [fetchDashboardData]);

  const rows: Rows = dashboardData.map(({ user, activeShunt }, id) => ({
    id: id,
    ...user,
    ...activeShunt,
  }));
  const columns = createColumns(rows);

  return (
    <Box sx={styles.box}>
      {/* Error Alert Banner */}
      <ErrorAlert
        error={error}
        onDismiss={() => setError(null)}
        onRetry={fetchDashboardData}
        retryButtonText="Retry"
      />

      {/* Loading Progress Bar */}
      {loading && <LinearProgress sx={{ mb: 2 }} />}

      {/* Action Buttons */}
      <Stack
        direction="row"
        spacing={2}
        alignItems="center"
        sx={{ mb: 2, mt: 2 }}
      >
        <Button
          variant="contained"
          sx={styles.button}
          onClick={() => setIsNewEntryOpen(true)}
        >
          {t("addEntry.button")}
        </Button>
        <IconButton
          onClick={fetchDashboardData}
          disabled={loading}
          sx={styles.refreshButton}
          aria-label="refresh data"
        >
          <RefreshIcon />
        </IconButton>
      </Stack>

      {/* Data Grid */}
      <StyledDashboardGrid
        columns={columns}
        rows={rows}
        pagination
        checkboxSelection
        disableRowSelectionOnClick
        loading={loading}
      />

      {/* New Entry Modal */}
      {isNewEntryOpen && (
        <NewEntry
          open={isNewEntryOpen}
          onClose={() => setIsNewEntryOpen(false)}
        />
      )}
    </Box>
  );
};

const styles = {
  box: {
    overflowX: "auto",
    maxWidth: "100%",
    display: "flex",
    flexDirection: "column",
    backgroundColor: "background.paper",
    pt: "64px",
    px: 2,
  },
  button: {
    p: 1,
  },
  refreshButton: {
    border: "1px solid",
    borderColor: "primary.main",
  },
};
export { Dashboard };
