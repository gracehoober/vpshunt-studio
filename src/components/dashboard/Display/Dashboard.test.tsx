import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { waitFor } from "@testing-library/react";

import { Dashboard } from "./Dashboard";

describe("<App>", () => {
  it("renders successfully", () => {
    render(<Dashboard />);
  });

  it("matches snapshot", () => {
    const { container } = render(<Dashboard />);
    expect(container).toMatchSnapshot();
  });

  it("contains the text 'Add Entry'", () => {
    const { getByText } = render(<Dashboard />);
    expect(getByText("addEntry.button")).toBeInTheDocument();
  });

  it("contains a button element", () => {
    const { container } = render(<Dashboard />);
    const button = container.querySelector("button");
    expect(button).toBeInTheDocument();
  });

  it("renders the StyledDashboardGrid component", () => {
    const { container } = render(<Dashboard />);
    const grid = container.querySelector(".MuiDataGrid-root");
    expect(grid).toBeInTheDocument();
  });

  it("opens NewEntry modal on button click", async () => {
    const { getByText, queryByLabelText } = render(<Dashboard />);
    const button = getByText("addEntry.button");
    expect(queryByLabelText("close")).not.toBeInTheDocument();
    button.click();
    await waitFor(() => {
      expect(queryByLabelText("close")).toBeInTheDocument();
    });
  });

  it("closes NewEntry modal on close button click", async () => {
    const { getByText, queryByLabelText, findByTestId } = render(<Dashboard />);
    const button = getByText("addEntry.button");
    button.click();
    const closeButton = await findByTestId("close-modal-button");
    expect(queryByLabelText("close")).toBeInTheDocument();
    closeButton.click();
    await waitFor(() => {
      expect(queryByLabelText("close")).not.toBeInTheDocument();
    });
  });
});
