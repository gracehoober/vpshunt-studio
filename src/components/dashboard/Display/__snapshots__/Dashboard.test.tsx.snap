// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<App> > matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-k6ke0i"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sctp6o-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      addEntry.button
    </button>
    <div
      class="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor MuiDataGridVariables-3851315942 css-10mzhy7-MuiDataGrid-root"
      style="--DataGrid-hasScrollX: 1; --DataGrid-hasScrollY: 0; --DataGrid-scrollbarSize: 0px; --DataGrid-rowWidth: 50px; --DataGrid-columnsTotalWidth: 50px; --DataGrid-leftPinnedWidth: 0px; --DataGrid-rightPinnedWidth: 0px; --DataGrid-headerHeight: 56px; --DataGrid-headersTotalHeight: 56px; --DataGrid-topContainerHeight: 56px; --DataGrid-bottomContainerHeight: 0px; --height: 52px;"
    >
      <div
        class="MuiDataGrid-mainContent"
        role="presentation"
      >
        <div
          aria-colcount="1"
          aria-multiselectable="true"
          aria-rowcount="1"
          class="MuiDataGrid-main css-1bgckiu-MuiDataGrid-main"
          role="grid"
          tabindex="-1"
        >
          <div
            class="css-18l11qw"
            data-id="gridPanelAnchor"
            role="presentation"
          />
          <div
            class="MuiDataGrid-virtualScroller MuiDataGrid-virtualScroller--hasScrollX css-1vigakz-MuiDataGrid-virtualScroller"
            role="presentation"
            style="overflow-x: hidden;"
          >
            <div
              class="MuiDataGrid-topContainer MuiDataGrid-container--top css-b1wygl"
              role="presentation"
            >
              <div
                class="MuiDataGrid-columnHeaders css-1gtv474-MuiDataGrid-columnHeaders"
                role="presentation"
              >
                <div
                  aria-rowindex="1"
                  class="MuiDataGrid-row--borderBottom css-yseucu-MuiDataGrid-columnHeaderRow"
                  role="row"
                  style="height: 56px;"
                >
                  <div
                    role="presentation"
                    style="width: 0px;"
                  />
                  <div
                    aria-colindex="1"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor MuiDataGrid-columnHeader--lastUnpinned MuiDataGrid-columnHeaderCheckbox MuiDataGrid-columnHeader--last"
                    data-field="__check__"
                    role="columnheader"
                    style="width: 50px;"
                    tabindex="0"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                      role="presentation"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                        role="presentation"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <span
                            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-2rjb2j-MuiButtonBase-root-MuiSwitchBase-root-MuiCheckbox-root"
                          >
                            <input
                              aria-label="Select all rows"
                              class="PrivateSwitchBase-input css-12xagqm-MuiSwitchBase-root"
                              data-indeterminate="false"
                              name="select_all_rows"
                              tabindex="-1"
                              type="checkbox"
                            />
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                              data-testid="CheckBoxOutlineBlankIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                              />
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-1umw9bq-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <rect
                          height="24"
                          rx="0.5"
                          width="1"
                          x="11.5"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-filler"
                    role="presentation"
                  />
                  <div
                    class="MuiDataGrid-scrollbarFiller MuiDataGrid-scrollbarFiller--header"
                    role="presentation"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiDataGrid-overlayWrapper css-107ft6p-MuiDataGrid-overlayWrapper"
            >
              <div
                class="MuiDataGrid-overlayWrapperInner css-1mbsuzk-MuiDataGrid-overlayWrapperInner"
                style="height: var(--DataGrid-overlayHeight, calc(var(--height) * 2)); width: 0px;"
              >
                <div
                  class="MuiDataGrid-overlay css-w16tsr-MuiDataGrid-overlay"
                >
                  No rows
                </div>
              </div>
            </div>
            <div
              class="MuiDataGrid-virtualScrollerContent css-aymtem-MuiDataGrid-virtualScrollerContent"
              role="presentation"
              style="width: auto; flex-basis: var(--DataGrid-overlayHeight, calc(var(--height) * 2)); flex-shrink: 0;"
            >
              <div
                class="MuiDataGrid-virtualScrollerRenderZone css-11dqcl8-MuiDataGrid-virtualScrollerRenderZone"
                role="rowgroup"
                style="transform: translate3d(0, 0px, 0);"
              />
            </div>
            <div
              class="MuiDataGrid-bottomContainer MuiDataGrid-container--bottom css-gzxaox"
              role="presentation"
            />
          </div>
          <div
            aria-hidden="true"
            class="MuiDataGrid-scrollbar MuiDataGrid-scrollbar--horizontal css-2n9dey"
            tabindex="-1"
          >
            <div
              class="MuiDataGrid-scrollbarContent"
            />
          </div>
        </div>
        <div
          class="MuiDataGrid-footerContainer MuiDataGrid-withBorderColor css-15lx25q-MuiDataGrid-footerContainer"
        >
          <div />
          <div
            class="MuiTablePagination-root css-y1ejz4-MuiTablePagination-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-1gak8h1-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                class="MuiTablePagination-spacer css-1wtxofq-MuiTablePagination-spacer"
              />
              <p
                class="MuiTablePagination-selectLabel css-s09cke-MuiTablePagination-selectLabel"
                id="«ra»"
              >
                Rows per page:
              </p>
              <div
                class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-select MuiSelect-root MuiTablePagination-input css-oegngy-MuiInputBase-root-MuiTablePagination-select"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="«ra» «r9»"
                  class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-18zjd9i-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input"
                  id="«r9»"
                  role="combobox"
                  tabindex="0"
                >
                  100
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="100"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-ug7h84-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                class="MuiTablePagination-displayedRows css-11cfq65-MuiTablePagination-displayedRows"
              >
                0–0 of 0
              </p>
              <div
                class="MuiTablePaginationActions-root MuiTablePagination-actions css-sjsoon-MuiTablePaginationActions-root"
              >
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-h7cjts-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to previous page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                    data-testid="KeyboardArrowLeftIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-h7cjts-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to next page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                    data-testid="KeyboardArrowRightIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <style
        href="/MuiDataGridVariables-3851315942"
      >
        .MuiDataGridVariables-3851315942{--DataGrid-t-spacing-unit:8px;--DataGrid-t-color-border-base:rgba(224, 224, 224, 1);--DataGrid-t-color-background-base:#fff;--DataGrid-t-color-background-overlay:#fff;--DataGrid-t-color-background-backdrop:rgba(255, 255, 255, 0.38);--DataGrid-t-color-foreground-base:rgba(0, 0, 0, 0.87);--DataGrid-t-color-foreground-muted:rgba(0, 0, 0, 0.6);--DataGrid-t-color-foreground-accent:#1565c0;--DataGrid-t-color-foreground-disabled:rgba(0, 0, 0, 0.38);--DataGrid-t-color-foreground-error:#c62828;--DataGrid-t-color-interactive-hover:rgba(0, 0, 0, 0.04);--DataGrid-t-color-interactive-hover-opacity:0.04;--DataGrid-t-color-interactive-focus:rgba(from #1976d2 r g b / 1);--DataGrid-t-color-interactive-focus-opacity:0.12;--DataGrid-t-color-interactive-disabled:rgba(from rgba(0, 0, 0, 0.26) r g b / 1);--DataGrid-t-color-interactive-disabled-opacity:0.38;--DataGrid-t-color-interactive-selected:#1976d2;--DataGrid-t-color-interactive-selected-opacity:0.08;--DataGrid-t-header-background-base:#fff;--DataGrid-t-cell-background-pinned:#fff;--DataGrid-t-radius-base:4px;--DataGrid-t-typography-font-family-base:"Roboto", "Helvetica", "Arial", sans-serif;--DataGrid-t-typography-font-weight-light:300;--DataGrid-t-typography-font-weight-regular:400;--DataGrid-t-typography-font-weight-medium:500;--DataGrid-t-typography-font-weight-bold:700;--DataGrid-t-typography-font-body:400 0.875rem / 1.43 "Roboto", "Helvetica", "Arial", sans-serif;--DataGrid-t-typography-font-small:400 0.75rem / 1.66 "Roboto", "Helvetica", "Arial", sans-serif;--DataGrid-t-typography-font-large:400 1rem / 1.5 "Roboto", "Helvetica", "Arial", sans-serif;--DataGrid-t-transition-easing-ease-in:cubic-bezier(0.4, 0, 1, 1);--DataGrid-t-transition-easing-ease-out:cubic-bezier(0.0, 0, 0.2, 1);--DataGrid-t-transition-easing-ease-in-out:cubic-bezier(0.4, 0, 0.2, 1);--DataGrid-t-transition-duration-short:200ms;--DataGrid-t-transition-duration-base:250ms;--DataGrid-t-transition-duration-long:300ms;--DataGrid-t-shadow-base:0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);--DataGrid-t-shadow-overlay:0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);--DataGrid-t-z-index-panel:1300;--DataGrid-t-z-index-menu:1300;}
      </style>
    </div>
  </div>
</div>
`;
